versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -0,0 +1,740 @@
+        Hash currentCommitId = head != null ? objIdToHash(head.id()) : NO_ANCESTOR;
+        verifyExpectedHash(currentCommitId, namedRef, expectedHash);
+        expected = reference(refName, hashToObjId(expectedHash.get()), false);
+      }
+
+      ObjId newPointer = hashToObjId(targetHash);
+      if (!EMPTY_OBJ_ID.equals(newPointer) && persist.fetchObjType(newPointer) != COMMIT) {
+        throw RefMapping.hashNotFound(targetHash);
+      }
+
+      referenceLogic.assignReference(expected, newPointer);
+    } catch (RefNotFoundException e) {
+      throw referenceNotFound(namedRef);
+    } catch (RefConditionFailedException e) {
+      throw referenceConflictException(
+          namedRef, objIdToHash(e.reference().pointer()), expected.pointer());
+    } catch (ObjNotFoundException e) {
+      throw referenceNotFound(e);
+    }
+  }
+
+  @Override
+  public Hash delete(NamedRef namedRef, Optional<Hash> hash)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    String refName = namedRefToRefName(namedRef);
+    ReferenceLogic referenceLogic = referenceLogic(persist);
+
+    ObjId expected = EMPTY_OBJ_ID;
+    try {
+      expected =
+          hash.isPresent()
+              ? hashToObjId(hash.get())
+              : referenceLogic.getReference(refName).pointer();
+      referenceLogic.deleteReference(refName, expected);
+      return objIdToHash(expected);
+    } catch (RefNotFoundException e) {
+      throw referenceNotFound(namedRef);
+    } catch (RefConditionFailedException e) {
+      RefMapping refMapping = new RefMapping(persist);
+      CommitObj headCommit = refMapping.resolveRefHead(namedRef);
+      throw referenceConflictException(
+          namedRef, objIdToHash(expected), headCommit != null ? headCommit.id() : EMPTY_OBJ_ID);
+    } catch (RetryTimeoutException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public ReferenceInfo<CommitMeta> getNamedRef(String refName, GetNamedRefsParams params)
+      throws ReferenceNotFoundException {
+    RefMapping refMapping = new RefMapping(persist);
+    Reference reference = refMapping.resolveNamedRef(refName);
+    NamedRef namedRef = referenceToNamedRef(reference);
+
+    CommitObj head = refMapping.resolveNamedRefHead(reference);
+
+    Supplier<CommitObj> refHeadSupplier = supplyRefHeadForNamedRefs(refMapping, params);
+
+    CommitLogic commitLogic = commitLogic(persist);
+    try {
+      return buildReferenceInfo(params, refHeadSupplier, commitLogic, namedRef, head);
+    } catch (ObjNotFoundException e) {
+      throw referenceNotFound(e);
+    }
+  }
+
+  private Supplier<CommitObj> supplyRefHeadForNamedRefs(
+      RefMapping refMapping, GetNamedRefsParams params) throws ReferenceNotFoundException {
+    NamedRef baseReference = params.getBaseReference();
+    if (baseReference == null) {
+      return () -> null;
+    }
+    Reference ref = refMapping.resolveNamedRef(baseReference);
```

Comment:
Why do we resolve the ref immediately, but load the commit in a supplier? What's the rationale for this distinction?

Reply:
yea, not really necessary, but also not overkill.. updated.